### PR TITLE
Fix kernel cache miss and add RDNA configs

### PR DESCRIPTION
**Pull request by hyoon1**
- added Navi configurations (Related PR: https://github.com/ROCm/triton/pull/640)
- resolved cache miss issue during flash attention calls by fixing max_seqlen_q/k to 0

### CODE REVIEW

File: vllm/attention/ops/triton_flash_attention.py [hunk 1]
```diff
@@ -795,8 +880,8 @@ def forward(
             HQ=nheads_q,
             HK=nheads_k,
             ACTUAL_BLOCK_DMODEL=head_size,
-            MAX_SEQLENS_Q=max_seqlens_q,
-            MAX_SEQLENS_K=max_seqlens_k,
+            MAX_SEQLENS_Q=0,
```
Comment (maleksan85) [inline vllm/attention/ops/triton_flash_attention.py:883]: what is the reason to zero seq lens?

File: vllm/attention/ops/triton_flash_attention.py [hunk 2]
```diff
@@ -207,103 +209,186 @@ def _attn_fwd_inner(
...
+def is_hip():
+    return triton.runtime.driver.active.get_current_target().backend == "hip"
+
+
+def is_cdna():
+    return is_hip() and triton.runtime.driver.active.get_current_target(
+    ).arch in ('gfx940', 'gfx941', 'gfx942', 'gfx90a', 'gfx908')
+
+
+def is_rdna():
```
Comment (maleksan85) [inline vllm/attention/ops/triton_flash_attention.py:241]: probably worth to use: https://github.com/ROCm/vllm/blob/8f3bf8b2b5bf1d231007629b3a5fbbf235f0cfe9/vllm/utils.py#L1620

File: vllm/attention/ops/triton_flash_attention.py [hunk 3]
```diff
@@ -207,103 +209,186 @@ def _attn_fwd_inner(
...
+    except Exception as e:
+        print(f"Error: {e}")
+    return None
+
+
+def is_hip():
+    return triton.runtime.driver.active.get_current_target().backend == "hip"
+
+
+def is_cdna():
```
Comment (maleksan85) [inline vllm/attention/ops/triton_flash_attention.py:236]: As per my knowledge AMD has two lines of HW for vllm: MI and Navi. So `not navi` should work better for future generations of MIs

**Comment by maleksan85** [comment]
As per @gshtras we need to merge into develop branch instead of main for now. Please correct.

File: vllm/attention/ops/triton_flash_attention.py [hunk 4]
```diff
@@ -207,103 +209,186 @@ def _attn_fwd_inner(
...
+            line = line.strip()
+            if line.startswith("Name: gfx"):
+                gfx_version = line.split("Name:")[1].strip()
+                return gfx_version
+    except Exception as e:
+        print(f"Error: {e}")
+    return None
+
+
+def is_hip():
```
Comment (gshtras) [inline vllm/attention/ops/triton_flash_attention.py:232]: All this functionality is implemented in a cross-architecture fashion in the platform/rocm.py and its superclasses

**Comment by hyoon1** [comment]
@maleksan85 @gshtras 
Thanks for the comments, I agree with your suggestions. However, I have a few concerns regarding this matter and I'm seeking advice on how to proceed. First, I believe that triton_flash_attention.py in vllm is essentially a copy of the file from [this link](https://github.com/ROCm/triton/blob/main_perf/python/perf-kernels/flash-attention.py). The modifications in this pull request directly apply the changes from the ROCm/triton repository's pull request [#640](https://github.com/ROCm/triton/pull/640). While it might be fine to change the relevant functions only in vllm, there is a risk of misalignment later on.

Secondly, our team is using the v0.6.2+rocm release, and I understand that functions like is_navi() are not supported in that version. Implementing them would require significant modifications. Therefore, maintaining backward compatibility is also a concern.

Given these considerations, I would greatly appreciate your advice on how to proceed with the modifications.

**Comment by gshtras** [comment]
As for your last point, whatever changes will be made here will not have any effect on the previous tags, so v0.6.2+rocm will not get affected.
That's true, our kernel is a snapshot of the one you mentioned, taken in the beginning of 2024. Our attempts to catch up in the past resulted in performance regressions on various models in different configs, which was noted to the team, but I don't believe it was thoroughly investigated. We may take another round of this experiment, but regardless, I think utilizing an existing infrastructure APIs that vllm provides is better here, if nothing else, for uniformity and avoiding code duplication.

File: vllm/attention/ops/triton_flash_attention.py (hunk 2 above)
Comment (hyoon1) [reply vllm/attention/ops/triton_flash_attention.py]: updated

File: vllm/attention/ops/triton_flash_attention.py (hunk 4 above)
Comment (hyoon1) [reply vllm/attention/ops/triton_flash_attention.py]: updated

File: vllm/attention/ops/triton_flash_attention.py (hunk 3 above)
Comment (hyoon1) [reply vllm/attention/ops/triton_flash_attention.py]: updated

File: vllm/attention/ops/triton_flash_attention.py [hunk 5]
```diff
@@ -207,103 +209,149 @@ def _attn_fwd_inner(
     return acc, l_i, m_i
 
 
-@triton.autotune(
-    configs=[
+def get_cdna_autotune_configs():
```
Comment (maleksan85) [inline vllm/attention/ops/triton_flash_attention.py:212]: are you sure that those new commits will not decrease performance on MI. If so, what models did you tested?
cc @gshtras

Comment (liangshen68) [reply vllm/attention/ops/triton_flash_attention.py:212]: We have tested chatglm2-6b, qwen-14b-chat, baichuan2-13b, llama-2-70b-chat, glm-4-9b-chat, qwen1.5-72b-chat-gptq, etc. on Navi31, w/o this change, triton-based FA2 has no positive perf lifting; while with this change, triton-based FA2 shows 2-5% gain. (and by debugging, it is confirmed that triton FA2 kernel cache is missed). We believe this should also provide positive impact on MI, especially during early triton kernel cache built-up period.

Comment (maleksan85) [reply vllm/attention/ops/triton_flash_attention.py:212]: discussed in the chat to separate things for MI from this PR.

Comment (hyoon1) [reply vllm/attention/ops/triton_flash_attention.py:212]: restored autotune configs for MI series

Comment (liangshen68) [reply vllm/attention/ops/triton_flash_attention.py:212]: @hyoon1, could you please make this change only applicable to Navi? I will ask engineers in China to confirm the perf gain on Navi32 (although such cache misses issue has no dependencies on what GPU used). Thanks.

Comment (hyoon1) [reply vllm/attention/ops/triton_flash_attention.py:212]: updated. MI will use original configs for autotune.

**Review by gshtras — APPROVED** [review]

File: vllm/attention/ops/triton_flash_attention.py (hunk 5 above)
Comment (hyoon1) [reply vllm/attention/ops/triton_flash_attention.py:212]: additional chatglm3-6b throughput test result on Navi 32 (16gb)
use triton / num-prompts 512 / max-model-len 512
original: input: 1234.33 toks/s, output: 921.11 toks/s Throughput: 5.31 requests/s, 2544.00 tokens/s
w/ update:  speed input: 1386.34 toks/s, output: 1034.54 toks/s Throughput: 5.96 requests/s, 2856.15 tokens/s

**Review by gshtras — DISMISSED** [review]

**Comment by gshtras** [comment]
Please try to avoid force pushes after the initial reviews. It makes it impossible to see the new changes.
Could you summarize what was changed since the original review?

File: vllm/attention/ops/triton_flash_attention.py [hunk 6]
```diff
@@ -314,8 +308,92 @@ def _attn_fwd_inner(
...
+        # Fall-back config.
+        triton.Config(
+            {
+                'BLOCK_M': 16,
+                'BLOCK_N': 16,
+                'waves_per_eu': 1,
+                'PRE_LOAD_V': False
+            },
+            num_stages=1,
+            num_warps=2),
```
Comment (ilia-cher) [inline vllm/attention/ops/triton_flash_attention.py:379]: we had to recently remove some of these configs - please check https://github.com/ROCm/vllm/pull/301 to make sure we don't put them back, otherwise we get Triton crashes for FP8 case

Comment (hyoon1) [reply vllm/attention/ops/triton_flash_attention.py:379]: Is it only related to BLOCK_M/BLOCK_N size? I see the num_warps are different. Should I remove all three configs which BLOCK_M=BLOCK_N=16?

Comment (ilia-cher) [reply vllm/attention/ops/triton_flash_attention.py:379]: it's related to the shapes, so BLOCK_M=BLOCK_N=16

Comment (hyoon1) [reply vllm/attention/ops/triton_flash_attention.py]: updated

**Review by gshtras — APPROVED** [review]

**Review by ilia-cher — APPROVED** [review]
